python/cudf_polars/cudf_polars/dsl/traversal.py

```diff
@@ -0,0 +1,191 @@
+    "make_recursive",
+    "CachingVisitor",
+]
+
+
+def traversal(node: Node) -> Generator[Node, None, None]:
+    """
+    Pre-order traversal of nodes in an expression.
+
+    Parameters
+    ----------
+    node
+        Root of expression to traverse.
+
+    Yields
+    ------
+    Unique nodes in the expression, parent before child, children
+    in-order from left to right.
+    """
+    seen = {node}
+    lifo = [node]
+
+    while lifo:
+        node = lifo.pop()
+        yield node
+        for child in reversed(node.children):
+            if child not in seen:
+                seen.add(child)
+                lifo.append(child)
+
+
+# reuse_if_unchanged can either be applied to expressions, in which
+# case we need an ExprTransformer...
+@overload
+def reuse_if_unchanged(node: expr.Expr, fn: ExprTransformer) -> expr.Expr: ...
+
+
+# .. or to plan nodes (IR), in which case we need an IRTransformer
+@overload
+def reuse_if_unchanged(node: ir.IR, fn: IRTransformer) -> ir.IR: ...
+
+
+def reuse_if_unchanged(node, fn):
+    """
+    Recipe for transforming nodes that returns the old object if unchanged.
+
+    Parameters
+    ----------
+    node
+         Node to recurse on
+    fn
+         Function to transform children
+
+    Notes
+    -----
+    This can be used as a generic "base case" handler when
+    writing transforms that take nodes and produce new nodes.
+
+    Returns
+    -------
+    Existing node `e` if transformed children are unchanged, otherwise
+    reconstructed node with new children.
+    """
+    new_children = [fn(c) for c in node.children]
+    if all(new == old for new, old in zip(new_children, node.children, strict=True)):
+        return node
+    return node.reconstruct(new_children)
+
+
+U_contra = TypeVar("U_contra", bound=Hashable, contravariant=True)
+V_co = TypeVar("V_co", covariant=True)
+
+
+def make_recursive(
+    fn: Callable[[U_contra, GenericTransformer[U_contra, V_co]], V_co],
+    *,
+    state: Mapping[str, Any] | None = None,
+) -> GenericTransformer[U_contra, V_co]:
+    """
+    No-op wrapper for recursive visitors.
+
+    Facilitates using visitors that don't need caching but are written
+    in the same style.
+
+    Parameters
+    ----------
+    fn
+        Function to transform inputs to outputs. Should take as its
+        second argument a callable from input to output.
+    state
+        Arbitrary *immutable* state that should be accessible to the
+        visitor through the `state` property.
+
+    Notes
+    -----
+    All transformation functions *must* be free of side-effects.
+
+    Usually, prefer a :class:`CachingVisitor`, but if we know that we
+    don't need caching in a transformation and then this no-op
+    approach is slightly cheaper.
+
+    Returns
+    -------
+    Recursive function without caching.
+
+    See Also
+    --------
+    CachingVisitor
+    """
+
+    def rec(node: U_contra) -> V_co:
+        return fn(node, rec)  # type: ignore[arg-type]
```

Comment:
I assume that it has no meaningful performance impact, but it might be worth adding a comment that this approach effectively doubles the size of the call stack. Aside from performance, it might make debuggging via stack traces a bit more confusing.

Reply:
It's marginally more expensive than recursing directly. I wrote a trivial test:

```
from functools import partial

def indirect(obj, rec):
    if isinstance(obj, int):
        return obj
    return sum(map(rec, obj))

def recursive(obj):
    return indirect(obj, recursive)

def direct(obj):
    if isinstance(obj, int):
        return obj
    return sum(map(direct, obj))

horrible = partial(indirect)
horrible.keywords["rec"] = horrible

tree = (1, 2, 3, (3, 4, 5, (6, 7, 8, (9, 10, 11, (12, 13, 14, (15, 16, 17))))))

tree = (1, 2, 3, (3, 4, 5, (6, 7, 8, (9, 10, 11, (12, 13, 14, (15, 16, 17, *tree))))))

tree = (1, 2, 3, (3, 4, 5, (6, 7, 8, (9, 10, 11, (12, 13, 14, (15, 16, 17, *tree))))))

tree = (1, 2, 3, (3, 4, 5, (6, 7, 8, (9, 10, 11, (12, 13, 14, (15, 16, 17, *tree))))))

%timeit direct(tree)
%timeit recursive(tree)
%timeit horrible(tree)
```

Results:
```
# 3.13
9.62 μs ± 37.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
11.9 μs ± 28.7 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
23.1 μs ± 657 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

# 3.12
10.6 μs ± 248 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
13.5 μs ± 297 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
23.1 μs ± 103 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

# 3.11
8.42 μs ± 135 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
11.2 μs ± 117 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
17.7 μs ± 59.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

# 3.10
11.2 μs ± 33.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
17.1 μs ± 27.6 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
21 μs ± 175 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

So not enough to be concerning. I am glad that the horrible approach is slower (even if it does give nicer stack traces).

The other approach is to just pass the state as an argument, rather than in the recursive `Transformer`. But then that means that any dag-aware traversal is intimately tied to the implementation, and you have to do it every time, by writing something like:

```
def rewrite(node: Node, state: Any, cache: dict[Node, T]) -> T:
    try:
        return cache[node]
    except KeyError:
        new_children = (rewrite(c, state, cache) for c in node.children)
        return cache.setdefault(node, node.reconstruct(new_children))
```
Which is painful and error-prone.

Reply:
Thanks for the benchmark, I'm fine with the current approach. I've done the direct passing of a cache down the stack before in other places, and I agree that it's inferior to your approach.